nullaway/src/main/java/com/uber/nullaway/dataflow/cfg/NullAwayCFGBuilder.java

```diff
@@ -0,0 +1,183 @@
+
+    private final Handler handler;
+
+    /**
+     * Create a new NullAway phase one translation visitor.
+     *
+     * @param builder a TreeBuilder object (used to create synthetic AST nodes to feed to the
+     *     translation process)
+     * @param annotationProvider an {@link AnnotationProvider}.
+     * @param assumeAssertionsEnabled can assertions be assumed to be disabled?
+     * @param assumeAssertionsDisabled can assertions be assumed to be enabled?
+     * @param env annotation processing environment containing type utilities
+     * @param handler a NullAway handler or chain of handlers (through {@link
+     *     com.uber.nullaway.handlers.CompositeHandler}
+     */
+    public NullAwayCFGTranslationPhaseOne(
+        TreeBuilder builder,
+        AnnotationProvider annotationProvider,
+        boolean assumeAssertionsEnabled,
+        boolean assumeAssertionsDisabled,
+        ProcessingEnvironment env,
+        Handler handler) {
+      super(builder, annotationProvider, assumeAssertionsEnabled, assumeAssertionsDisabled, env);
+      this.handler = handler;
+    }
+
+    @SuppressWarnings("NullAway") // (Void)null issue
+    private void scanWithVoid(Tree tree) {
+      this.scan(tree, (Void) null);
+    }
+
+    /**
+     * Obtain the type mirror for a given class, used for exception throwing.
+     *
+     * @param klass a Java class
+     * @return the corresponding type mirror
+     */
+    public TypeMirror classToErrorType(Class<?> klass) {
+      return this.getTypeMirror(klass);
+    }
+
+    /**
+     * Extend the CFG to throw an exception if the passed expression node evaluates to false.
+     *
+     * @param booleanExpressionNode a CFG Node representing a boolean expression.
+     * @param errorType the type of the exception to throw if booleanExpressionNode evaluates to
+     *     false.
+     */
+    public void insertThrowOnFalse(Node booleanExpressionNode, TypeMirror errorType) {
+      Tree tree = booleanExpressionNode.getTree();
+      Preconditions.checkArgument(
+          tree instanceof ExpressionTree,
+          "Argument booleanExpressionNode must represent a boolean expression");
+      ExpressionTree booleanExpressionTree = (ExpressionTree) booleanExpressionNode.getTree();
+      Preconditions.checkNotNull(booleanExpressionTree);
+      Label falsePreconditionEntry = new Label();
+      Label endPrecondition = new Label();
+      this.scanWithVoid(booleanExpressionTree);
+      ConditionalJump cjump = new ConditionalJump(endPrecondition, falsePreconditionEntry);
+      this.extendWithExtendedNode(cjump);
+      this.addLabelForNextNode(falsePreconditionEntry);
+      ExtendedNode exNode =
+          this.extendWithNodeWithException(
+              new ThrowNode(
+                  new ThrowTree() {
+                    @Override
+                    public ExpressionTree getExpression() {
+                      return booleanExpressionTree;
```

Comment:
Not sure if it matters, but this is not quite right, yes?  To be valid I think this should be a tree for `new Exception()` or something like that.  Of course creating the tree for the `new Exception()` expression would be some extra code.  This probably doesn't matter since I don't think NullAway ever relies on the exact exception value being thrown.

Reply:
The problem is also that this stuff is constantly being casted to `JCTree` instances, so the more deeply we generate "fake" `Tree`s, the higher the chance some parsing in the CFG conversion complains...

Reply:
FWIW, existing code in `CFGTranslationPhaseOne` just passes `null` as the `Tree`:

https://github.com/typetools/checker-framework/blob/ea1ff6d87b4e6f4bbf89c7eb4ecdf1769403e1a0/dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java#L1581-L1583

This won't pass NullAway checking, but we could get around that with a suppression.  Not sure if it's any better than building a dummy tree.  At least it indicates that however this dummy tree is constructed shouldn't matter _too_ much.

Reply:
Added a comment. Still feel better about the dummy tree than a NullAway suppression, but I admit there is no good reason for that preference 🤷

Reply:
SGTM